sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs

```
@@ -0,0 +1,99 @@
+
+            string remoteFetchError = string.Empty;
+            while (work.Count != 0 && !cancellationToken.IsCancellationRequested)
+            {
+                string tryContentPath = work.Dequeue();
+                ResolverEventSource.Shared.FetchingModelContent(tryContentPath);
+
+                string content = await EvaluatePathAsync(tryContentPath, cancellationToken).ConfigureAwait(false);
+                if (!string.IsNullOrEmpty(content))
+                {
+                    return new FetchResult()
+                    {
+                        Definition = content,
+                        Path = tryContentPath
+                    };
+                }
+
+                ResolverEventSource.Shared.ErrorFetchingModelContent(tryContentPath);
+                remoteFetchError = string.Format(CultureInfo.InvariantCulture, StandardStrings.ErrorFetchingModelContent, tryContentPath);
+            }
+
+            throw new RequestFailedException(remoteFetchError);
+        }
+
+        private static string GetPath(string dtmi, Uri repositoryUri, bool expanded = false)
+        {
+            string absoluteUri = repositoryUri.AbsoluteUri;
+            return DtmiConventions.DtmiToQualifiedPath(dtmi, absoluteUri, expanded);
+        }
+
+        private async Task<string> EvaluatePathAsync(string path, CancellationToken cancellationToken)
+        {
+            Request request = _pipeline.CreateRequest();
+            request.Method = RequestMethod.Get;
+            request.Uri = new RequestUriBuilder();
+            request.Uri.Reset(new Uri(path));
+
+            Response response = await _pipeline.SendRequestAsync(request, cancellationToken).ConfigureAwait(false);
+            if (response.Status >= 200 && response.Status <= 299)
+            {
+                return await GetContentAsync(response.ContentStream, cancellationToken).ConfigureAwait(false);
+            }
+
+            return null;
```

Why not surface an exception with error details?

---

well, I think the SendRequestAsync will throw if there is an exception and it will surface, the method can use some refactoring though

---

task added

---

If you look at the generated autorest code, you'll see differently. We need to throw a specific type of exception.

However, if we make swagger and generate, we'll get all this for free.